src/vector_commitment_of_discrete_log.rs

```diff
@@ -0,0 +1,691 @@
+// Author: dWallet Labs, Ltd.
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+use std::{marker::PhantomData, ops::Mul};
+
+use commitment::{GroupsPublicParametersAccessors, HomomorphicCommitmentScheme};
+use group::helpers::FlatMapResults;
+use group::{direct_product, self_product, BoundedGroupElement, CyclicGroupElement, Samplable};
+use serde::{Deserialize, Serialize};
+
+use crate::{language::GroupsPublicParameters, Error, Result, SOUND_PROOFS_REPETITIONS};
+
+/// Vector Commitment of Discrete Log Maurer Language:
+/// $$ (\vec{m}, \vec{\rho}) \mapsto (C_{\mainIndex}=\Com_\pp(m_{\mainIndex}; \rho_{\mainIndex}) \wedge
+///                     Y = \sum_\mainIndex m_\mainIndex\cdot G_\mainIndex) $$
+/// SECURITY NOTICE:
+/// Because correctness and zero-knowledge is guaranteed for any group in this language, we choose
+/// to provide a fully generic implementation.
+///
+/// However knowledge-soundness proofs are group dependent, and thus we can only assure security for
+/// groups for which we know how to prove it.
+///
+/// In the paper, we have proved it for any prime known-order group; so it is safe to use with a
+/// `PrimeOrderGroupElement`.
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug, Eq)]
+pub struct Language<
+    const BATCH_SIZE: usize,
+    const SCALAR_LIMBS: usize,
+    Scalar,
+    GroupElement,
+    CommitmentScheme,
+> {
+    _scalar_choice: PhantomData<Scalar>,
+    _group_element_choice: PhantomData<GroupElement>,
+    _commitment_choice: PhantomData<CommitmentScheme>,
+}
+
+impl<
+        const BATCH_SIZE: usize,
+        const SCALAR_LIMBS: usize,
+        Scalar: BoundedGroupElement<SCALAR_LIMBS>
+            + Samplable
+            + Mul<GroupElement, Output = GroupElement>
+            + for<'r> Mul<&'r GroupElement, Output = GroupElement>
+            + Copy,
+        GroupElement: CyclicGroupElement,
+        CommitmentScheme: HomomorphicCommitmentScheme<
+            SCALAR_LIMBS,
+            MessageSpaceGroupElement = self_product::GroupElement<BATCH_SIZE, Scalar>,
+        >,
+    > crate::Language<SOUND_PROOFS_REPETITIONS>
+    for Language<BATCH_SIZE, SCALAR_LIMBS, Scalar, GroupElement, CommitmentScheme>
+{
+    type WitnessSpaceGroupElement = direct_product::GroupElement<
```

Comment:
Another way to maybe say it better is that you can work with any commitment scheme with a message space that can be interpreted as a direct product of the scalar. For example this can be a vector containing three elements: A multi-pedersen committing to 5 elements, a regular Pedersen committing to a single value and an ElGamal commitment. Overall this new commitment can be thought of as the scalars to the power of 7 as a direct product.

Reply:
But this is what I did:
`            MessageSpaceGroupElement = self_product::GroupElement<BATCH_SIZE, Scalar>,`